File: DailyCore/Sources/DailyCore/Services/WpJsonParser.swift
import Foundation

/// Parser for WordPress REST API posts (`/wp-json/wp/v2/posts?_embed`).
public final class WpJsonParser: @unchecked Sendable {
    
    private let feed: FeedSource
    
    public init(feed: FeedSource) {
        self.feed = feed
    }
    
    public func parse(jsonData: Data) -> [NewsArticle] {
        guard let jsonArray = try? JSONSerialization.jsonObject(with: jsonData) as? [[String: Any]] else {
            return []
        }
        
        var articles: [NewsArticle] = []
        let isoFormatter = ISO8601DateFormatter()
        let fallbackFormatter = DateFormatter()
        fallbackFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        
        for post in jsonArray {
            let titleObj = post["title"] as? [String: Any]
            var title = titleObj?["rendered"] as? String ?? "No Title"
            title = decodeHtmlEntities(title)
            
            let link = post["link"] as? String ?? ""
            let dateStr = post["date"] as? String ?? ""
            let date = isoFormatter.date(from: dateStr) ?? fallbackFormatter.date(from: dateStr) ?? Date()
            
            let contentObj = post["content"] as? [String: Any]
            let content = contentObj?["rendered"] as? String
            
            let excerptObj = post["excerpt"] as? [String: Any]
            let excerpt = excerptObj?["rendered"] as? String
            
            var cleanDesc = stripHtmlTags(excerpt ?? "")
            if cleanDesc.isEmpty && content != nil {
                let contentClean = stripHtmlTags(content ?? "")
                cleanDesc = contentClean.count > 280 ? String(contentClean.prefix(280)) + "..." : contentClean
            }
            
            var authorName: String?
            var imageUrl: String?
            
            if let embedded = post["_embedded"] as? [String: Any] {
                // Author
                if let authors = embedded["author"] as? [[String: Any]], let firstAuthor = authors.first {
                    authorName = firstAuthor["name"] as? String
                }
                
                // Featured Media
                if let mediaList = embedded["wp:featuredmedia"] as? [[String: Any]], let firstMedia = mediaList.first {
                    imageUrl = firstMedia["source_url"] as? String
                    if let details = firstMedia["media_details"] as? [String: Any],
                       let sizes = details["sizes"] as? [String: Any],
                       let medium = sizes["medium"] as? [String: Any],
                       let mediumUrl = medium["source_url"] as? String {
                        imageUrl = mediumUrl
                    }
                }
            }
            
            articles.append(NewsArticle(
                id: link.isEmpty ? UUID().uuidString : link,
                title: title,
                link: link,
                publishDate: date,
                imageUrl: imageUrl ?? feed.iconUrl,
                description: cleanDesc,
                content: content,
                author: authorName,
                publicationName: feed.name,
                publicationIconUrl: feed.iconUrl,
                category: feed.category
            ))
        }
        
        return articles
    }
    
    private func stripHtmlTags(_ str: String) -> String {
        var clean = str
        clean = clean.replacingOccurrences(of: "</p>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br/>", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<br />", with: " ", options: .caseInsensitive)
        clean = clean.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression)
        clean = decodeHtmlEntities(clean)
        clean = clean.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        return clean.trimmingCharacters(in: .whitespacesAndNewlines)
    }
    
    private func decodeHtmlEntities(_ str: String) -> String {
        var decoded = str
        let entities = [
            ("&quot;", "\""),
            ("&apos;", "'"),
            ("&amp;", "&"),
            ("&lt;", "<"),
            ("&gt;", ">"),
            ("&nbsp;", " ")
        ]
        for (ent, val) in entities {
            decoded = decoded.replacingOccurrences(of: ent, with: val)
        }
        
        // Handle decimal numeric entities: &#539;
        if let regex = try? NSRegularExpression(pattern: "&#(\\d+);") {
            let ns = decoded as NSString
            let matches = regex.matches(in: decoded, range: NSRange(location: 0, length: ns.length)).reversed()
            var mod = decoded
            for m in matches {
                let numStr = ns.substring(with: m.range(at: 1))
                if let code = UInt32(numStr), let scalar = UnicodeScalar(code) {
                    if let targetRange = Range(m.range, in: mod) {
                        mod.replaceSubrange(targetRange, with: String(scalar))
                    }
                }
            }
            decoded = mod
        }
        
        // Handle hex numeric entities: &#x21b;
        if let regex = try? NSRegularExpression(pattern: "&#[xX]([0-9a-fA-F]+);") {
            let ns = decoded as NSString
            let matches = regex.matches(in: decoded, range: NSRange(location: 0, length: ns.length)).reversed()
            var mod = decoded
            for m in matches {
                let hexStr = ns.substring(with: m.range(at: 1))
                if let code = UInt32(hexStr, radix: 16), let scalar = UnicodeScalar(code) {
                    if let targetRange = Range(m.range, in: mod) {
                        mod.replaceSubrange(targetRange, with: String(scalar))
                    }
                }
            }
            decoded = mod
        }
        
        return decoded
    }
}

File: Android/core-model/src/main/java/com/intellidream/daily/model/WpJsonParser.kt
package com.intellidream.daily.model

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

class WpJsonParser(private val feed: FeedSource) {

    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    fun parse(jsonString: String): List<NewsArticle> {
        val articles = mutableListOf<NewsArticle>()
        try {
            val element = json.parseToJsonElement(jsonString)
            if (element is JsonArray) {
                for (item in element) {
                    if (item is JsonObject) {
                        val article = parsePost(item)
                        if (article != null) {
                            articles.add(article)
                        }
                    }
                }
            }
        } catch (_: Exception) {}
        return articles
    }

    private fun parsePost(post: JsonObject): NewsArticle? {
        val link = post["link"]?.jsonPrimitive?.content ?: ""
        val titleObj = post["title"]?.jsonObject
        val rawTitle = titleObj?.get("rendered")?.jsonPrimitive?.content ?: "No Title"
        val title = FeedParser.decodeHtmlEntities(rawTitle)

        val dateStr = post["date"]?.jsonPrimitive?.content ?: ""
        val date = parseDate(dateStr)

        val contentObj = post["content"]?.jsonObject
        val content = contentObj?.get("rendered")?.jsonPrimitive?.content

        val excerptObj = post["excerpt"]?.jsonObject
        val excerpt = excerptObj?.get("rendered")?.jsonPrimitive?.content ?: ""

        var cleanDesc = FeedParser.stripHtmlTags(excerpt)
        if (cleanDesc.isEmpty() && !content.isNullOrEmpty()) {
            val contentClean = FeedParser.stripHtmlTags(content)
            cleanDesc = if (contentClean.length > 280) contentClean.take(280) + "..." else contentClean
        }

        var authorName: String? = null
        var imageUrl: String? = null

        val embedded = post["_embedded"]?.jsonObject
        if (embedded != null) {
            val authors = embedded["author"]?.jsonArray
            if (authors != null && authors.isNotEmpty()) {
                val firstAuthor = authors[0].jsonObject
                authorName = firstAuthor["name"]?.jsonPrimitive?.content
            }

            val mediaList = embedded["wp:featuredmedia"]?.jsonArray
            if (mediaList != null && mediaList.isNotEmpty()) {
                val firstMedia = mediaList[0].jsonObject
                imageUrl = firstMedia["source_url"]?.jsonPrimitive?.content
                val details = firstMedia["media_details"]?.jsonObject
                val sizes = details?.get("sizes")?.jsonObject
                val medium = sizes?.get("medium")?.jsonObject
                val mediumUrl = medium?.get("source_url")?.jsonPrimitive?.content
                if (!mediumUrl.isNullOrEmpty()) {
                    imageUrl = mediumUrl
                }
            }
        }

        return NewsArticle(
            id = if (link.isNotBlank()) link else UUID.randomUUID().toString(),
            title = title,
            link = link,
            publishDate = date,
            imageUrl = imageUrl ?: feed.iconUrl,
            description = cleanDesc,
            content = content,
            author = authorName,
            publicationName = feed.name,
            publicationIconUrl = feed.iconUrl,
            category = feed.category
        )
    }

    private fun parseDate(str: String): Long {
        if (str.isBlank()) return System.currentTimeMillis()

        val formats = listOf(
            "yyyy-MM-dd'T'HH:mm:ss",
            "yyyy-MM-dd'T'HH:mm:ssZ",
            "yyyy-MM-dd'T'HH:mm:ssXXX"
        )
        for (fmt in formats) {
            try {
                val sdf = SimpleDateFormat(fmt, Locale.US)
                sdf.timeZone = TimeZone.getTimeZone("UTC")
                val d = sdf.parse(str)
                if (d != null) return d.time
            } catch (_: Exception) {}
        }

        try {
            return java.time.Instant.parse(str).toEpochMilli()
        } catch (_: Exception) {}

        return System.currentTimeMillis()
    }
}
